heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -7889,6 +7890,21 @@ proveExVarImpl :: NuMatchingAny1 r => PartialSubst vars -> Mb vars (Name tp) ->
                   Mb vars (ValuePerm tp) ->
                   ImplM vars s r (ps :> tp) ps (Name tp)
 
+-- If the variable x is a free variable of type unit, and if there is a global
+-- unit u in the context, instanitate x to u and call proveVarImpl
+proveExVarImpl _psubst mb_x mb_p
+  | Right n <- mbNameBoundP mb_x
+  , Just  u <- getUnitImplM
```

Comment:
Instead of `Just u <- getUnitImplM` this should be something like `u <- ensureUnitImplM`, where the `ensure` function will create a unit variable if one does not yet exist.

Reply:
Just realized this block doesn't even check to see whether `tp == Unit`...

Reply:
Removed this block altogether.